### PR TITLE
fix: changing plans

### PR DESCRIPTION
**Pull request by johncorrelli**
wip

### CODE REVIEW

File: src/Arr.php [hunk 1]
```diff
@@ -133,4 +133,14 @@ public static function assocIn(array $array, array $path, $value)
     {
         return static::updateIn($array, $path, fn () => $value);
     }
+
+    /**
+     * @param mixed[] $array
+     *
+     * @return mixed[]
+     */
+    public static function removeAndReIndex(array $array, callable $fn): array
```
Comment (brianium) [inline src/Arr.php:142]: We should just call this `filter`. `Arr::filter`'s claim to fame will be that it doesn't return sparse keys 💯

File: src/Processor/Braintree/Mapper/Subscription/ModifierMapper.php [hunk 2]
```diff
@@ -85,6 +85,11 @@ private function withRemovedPreviousPlanDefaults(array &$request, array $default
                 $request['remove'] = [];
             }
             $request['remove'][] = $default->getId();
+
+            // removed Modifiers can't be in the update
+            if (isset($request['update'])) {
```
Comment (brianium) [inline src/Processor/Braintree/Mapper/Subscription/ModifierMapper.php:95]: I knew this was coming... It just makes TOO MUCH SENSE

File: src/Processor/Braintree/Mapper/Subscription/ModifierMapper.php [hunk 3]
```diff
@@ -53,6 +53,11 @@ private function withNewDefaultModifiers(array &$request, array $newModifiers):
                 $default = Arr::replaceKeys($newDefault->toArray(), ['id' => 'inheritedFromId']);
 
                 if (is_array($modifiers)) {
+                    // if new modifier is in default, merge modifiers
+                    if ($modifiers[0]['inheritedFromId'] === $default['inheritedFromId']) {
```
Comment (brianium) [inline src/Processor/Braintree/Mapper/Subscription/ModifierMapper.php:57]: I don't understand this `[0]` - what if there are multiple modifiers?

File: src/Processor/Braintree/Subscription/Update/BaseUpdateStrategy.php [hunk 4]
```diff
@@ -43,6 +43,8 @@ protected function doBraintreeUpdate(Subscription $subscription, ?Plan $newPlan
         $request = Arr::dissoc($request, ['firstBillingDate', 'status']);
         $request['options'] = ['prorateCharges' => true];
 
+        // print_r($request);
```
Comment (brianium) [inline src/Processor/Braintree/Subscription/Update/BaseUpdateStrategy.php:46]: 😱

File: src/Processor/Braintree/Mapper/Subscription/ModifierMapper.php [hunk 5]
```diff
@@ -112,8 +117,17 @@ private function toModifiersRequest(Modifiers $modifiers, Plan $plan): array
             $modifier = $operation->getModifier();
             $type = $operation->getType();
             $default = $plan->getModifier($modifier->getId());
+            $isAdding = null === $default;
+
+            // when changing plans, allow modifications to default AddOns
+            if (!$subscription->isNew()) {
+                $defaultAddOns = array_keys($subscription->getPlan()->getAddOns());
```
Comment (brianium) [inline src/Processor/Braintree/Mapper/Subscription/ModifierMapper.php:124]: Anything with the `HasModifiers` trait can call `getAddOn` or `getDiscount` with an id. The `array_keys`/`in_array` approach seems unnecessary. Essentially:

```php
$defaultAddOns = $subscription->getPlan()->getAddOns();
$isDefaultAddOn = isset($defaultAddOns[$modifier->getId()]);
```

It also looks like this needs to account for discounts.

**Review by brianium — CHANGES_REQUESTED** [review]

File: src/Arr.php (hunk 1 above)
Comment (johncorrelli) [reply src/Arr.php:142]: I was struggling to come up with a good name for this... I like that idea... our helper does the magic! :handshake:

File: src/Processor/Braintree/Subscription/Update/BaseUpdateStrategy.php (hunk 4 above)
Comment (johncorrelli) [reply src/Processor/Braintree/Subscription/Update/BaseUpdateStrategy.php:46]: :table-flip: - I guess this PR is still wip, lol

**Review by brianium — APPROVED** [review]

**Pull request merged**